File: CPP/binary_tree_preorder_traversal.cpp
class Solution {
public:
    vector<int> preorderTraversal(TreeNode *root) {
        stack<TreeNode*> tn_stack;
        if (root != NULL) {
            tn_stack.push(root);
        }
        
        vector<int> result;
        while (!tn_stack.empty()) {
            root = tn_stack.top();
            tn_stack.pop();
            result.push_back(root->val);
            if (root->right != NULL) {
                tn_stack.push(root->right);
            }
            if (root->left != NULL) {
                tn_stack.push(root->left);
            }
            
        }
        return result;
    }
};

File: JAVA/BinaryTreePreorderTraversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> result = new LinkedList<Integer>();
        if (root != null) {
            Stack<TreeNode> stack = new Stack<TreeNode>();
            stack.push(root);
            while(!stack.isEmpty()) {
                root = stack.pop();
                result.add(root.val);
                if (root.right != null)
                    stack.push(root.right);
                if (root.left != null)
                    stack.push(root.left);
            }
        }
        
        return result;
    }
}